### PR TITLE
add name to service key in auto allocation

### PR DESCRIPTION
**Pull request by ramaraochavali**
**Please provide a description of this PR:**

### CODE REVIEW

File: pilot/pkg/serviceregistry/serviceentry/controller.go [hunk 1]
```diff
@@ -969,7 +969,7 @@ func autoAllocateIPs(services []*model.Service) []*model.Service {
 }
 
 func makeServiceKey(svc *model.Service) string {
-	return svc.Attributes.Namespace + "/" + svc.Hostname.String()
+	return svc.Attributes.Name + "/" + svc.Attributes.Namespace + "/" + svc.Hostname.String()
```
[the remark this answers is not in the document]
Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/serviceentry/controller.go]: Weird, for k8s service, this name equals to svc name

[the remark this answers is not in the document]
Comment (ramaraochavali) [reply pilot/pkg/serviceregistry/serviceentry/controller.go]: https://github.com/istio/istio/blob/321ba12d16d1d9746b6b9262fbbdba8c8deafa94/pilot/pkg/serviceregistry/serviceentry/conversion.go#L228

**Review by costinm — COMMENTED** [review]
Looks good - but looking at the tests I don't see how adding the name results in IP changes but the new name doesn't show up :-)

File: pilot/pkg/serviceregistry/serviceentry/controller.go [hunk 2]
```diff
@@ -969,7 +969,7 @@ func autoAllocateIPs(services []*model.Service) []*model.Service {
 }
 
 func makeServiceKey(svc *model.Service) string {
-	return svc.Attributes.Namespace + "/" + svc.Hostname.String()
+	return svc.Attributes.ConfigName + "/" + svc.Attributes.Namespace + "/" + svc.Hostname.String()
```
[the remark this answers is not in the document]
Comment (costinm) [reply pilot/pkg/serviceregistry/serviceentry/controller.go:972]: I mean - would be good to add a comment ( hashing , etc) - the code is pretty tricky to understand without comments. 

What would be even better is a detailed doc on how auto-allocation work, in the docs dir ( if there is one - I missed it ), and perhaps a section on how we see the evolution of this in a world of external dns, ambient, etc.

**Comment by hzxuzhonghu** [comment]
I am afraid this will not take effect for https://github.com/istio/istio/issues/48055

here n = makeServiceKey(svc), so same {namespace, hostname} already allocated same ip

```
		if v, ok := hnMap[n]; ok {
			setAutoAllocatedIPs(svc, v)
		} 
```

**Comment by ramaraochavali** [comment]
> here n = makeServiceKey(svc), so same {namespace, hostname} already allocated same ip

Yes. In fact this PR is incorrect as we can not have different IP for same host

**Comment by hzxuzhonghu** [comment]
And i did some dig on this function, found many redundant check. Will first cleanup them, mostly because we introduced maxIP, which is a safeguard

**Comment by ramaraochavali** [comment]
>  mostly because we introduced maxIP, which is a safeguard

I think it is because 240.240.0.0/16 can not have more than 65536 ips

**Pull request closed without merging**